crates/iota-transaction-builder/src/types/move_arg.rs

```diff
@@ -1,104 +1,113 @@
-    fn arg_type(&self) -> ArgType {
-        let mut res = u32_as_uleb128(self.len() as u32);
-        for val in self {
-            match val.arg_type() {
-                ArgType::Object(object_id) => res.extend(object_id.as_bytes()),
-                ArgType::Pure(items) => res.extend(items),
-            }
-        }
-        ArgType::Pure(res)
+/// A trait which defines how collections of move arg types are serialized for
+/// move calls.
+pub trait MoveArgCollection {
+    /// Get the pure BCS bytes.
+    fn collection_bytes(self) -> PureBytes;
+}
+
+impl<T: MoveArg> MoveArgCollection for T {
+    fn collection_bytes(self) -> PureBytes {
+        self.pure_bytes()
     }
 }
 
-fn u32_as_uleb128(mut value: u32) -> Vec<u8> {
-    let mut res = Vec::new();
-    while value >= 0x80 {
-        // Write 7 (lowest) bits of data and set the 8th bit to 1.
-        let byte = (value & 0x7f) as u8;
-        res.push(byte | 0x80);
-        value >>= 7;
+impl<const N: usize, T: MoveArg> MoveArgCollection for [T; N] {
+    fn collection_bytes(self) -> PureBytes {
+        PureBytes(
+            u32_as_uleb128(self.len() as u32)
+                .into_iter()
+                .chain(self.into_iter().flat_map(|val| val.pure_bytes().0))
+                .collect(),
+        )
     }
-    // Write the remaining bits of data and set the highest bit to 0.
-    res.push(value as u8);
-    res
 }
 
-impl<T: MoveArg> MoveArg for Box<T> {
-    fn arg_type(&self) -> ArgType {
-        self.as_ref().arg_type()
+impl<T: MoveArg> MoveArgCollection for Vec<T> {
+    fn collection_bytes(self) -> PureBytes {
+        PureBytes(
+            u32_as_uleb128(self.len() as u32)
+                .into_iter()
+                .chain(self.into_iter().flat_map(|val| val.pure_bytes().0))
+                .collect(),
+        )
     }
 }
 
-impl<T: MoveArg> MoveArg for Option<T> {
-    fn arg_type(&self) -> ArgType {
+impl<T: MoveArg> MoveArgCollection for Option<T> {
+    fn collection_bytes(self) -> PureBytes {
         match self {
-            Some(value) => match value.arg_type() {
-                ArgType::Object(object_id) => ArgType::Pure([&[1], object_id.as_bytes()].concat()),
-                ArgType::Pure(items) => ArgType::Pure([&[1], &items[..]].concat()),
-            },
-            None => ArgType::Pure(vec![0; 1]),
+            Some(value) => PureBytes([&[1], &value.pure_bytes().0[..]].concat()),
+            None => PureBytes(vec![0; 1]),
         }
     }
 }
 
-impl<T: MoveArg> MoveArg for &T {
-    fn arg_type(&self) -> ArgType {
-        (*self).arg_type()
+fn u32_as_uleb128(mut value: u32) -> Vec<u8> {
+    let mut res = Vec::new();
+    while value >= 0x80 {
+        // Write 7 (lowest) bits of data and set the 8th bit to 1.
+        let byte = (value & 0x7f) as u8;
+        res.push(byte | 0x80);
+        value >>= 7;
     }
+    // Write the remaining bits of data and set the highest bit to 0.
+    res.push(value as u8);
+    res
 }
```

Comment:
do we have tests for this? Why do we have to do bit manipulation code ourselves? Isn't there some battle tested crate that does this already? Or was this just moved from somewhere?